docs/configuration.md
# Configuration

The backend and client must be configured via environment variables. Some configuration-properties are mandatory and must be set to a correct/valid value, others are optional.

## Backend

All available configuration-properties are described directly inside the `config.js` file or in [Backend Code/config](Backend_code/config) (generated from source).

To configure the backend you need to edit the environment variables in your `docker-compose.yml`:

```yaml
services:
  # ...
  table-backend:
    # ...
    environment:
      # ...
      TABLE_ADMIN_PASSWORD: YOU_MUST_CHANGE_THIS_VALUE # change this to a long secret phrase
      TABLE_BASE_URL: https://table.example.com # url to client
      TABLE_DB_URL: mongodb://mongo:27017 # url to mongo instance
      TABLE_LDAP_URL: ldaps://ldap.example.com:636 # url to ldap server
      TABLE_LDAP_DN: uid=example,ou=example,dc=example,dc=com # login to search account
      TABLE_LDAP_PASSWORD: 12345678 # login to search account
      TABLE_LDAP_BASE: ou=people,dc=example,dc=com
      TABLE_MAIL_HOST: mailserver.example.com # mail-server
      TABLE_MAIL_FROM: no-reply@table.example.com # email address for automatic mails / notifications
      TABLE_SESSION_TOKEN_PRIVATE_KEY: secret-token-for-encryption # set this to a random value
```


## (Web-)Client

All available configuration-properties are described directly inside the `config.js` file or in [Client Code/config](Client_code/config) (generated from source).

To configure the backend you need to edit the environment variables in your `docker-compose.yml`:

```yaml
services:
  # ...
  table-client:
    # ...
    environment:
      # ...
      TABLE_API_URL: https://table-api.example.com # full url to backend
      TABLE_HTML_CONTACT_INFOS: |
        <p><strong>My Name</strong><br />Address and Tel-Nr.</p>
      TABLE_HTML_TERMS_OF_SERVICE: |
        <h2>lorem ipsum<h2>
        <p>lorem ipsum</p>
```


docs/Client_code/config.md
<a id="config"></a>

## config
Configuration values.
IMPORTANT: Only string values are supported!

**Kind**: global constant  

* [config](#config)
    * [.TABLE_API_DISCONNECT_TIMEOUT](#config46table95api95disconnect95timeout) : <code>string</code>
    * [.TABLE_API_REQUEST_TIMEOUT](#config46table95api95request95timeout) : <code>string</code>
    * [.TABLE_API_URL](#config46table95api95url) : <code>string</code>
    * [.TABLE_DESKTOP_APP_BROADCAST_IMAGE_INTERVAL](#config46table95desktop95app95broadcast95image95interval) : <code>string</code>
    * [.TABLE_DESKTOP_APP_SCREENSHOT_MAX_RES](#config46table95desktop95app95screenshot95max95res) : <code>string</code>
    * [.TABLE_HTML_CONTACT_INFOS](#config46table95html95contact95infos) : <code>string</code>
    * [.TABLE_HTML_TERMS_OF_SERVICE](#config46table95html95terms95of95service) : <code>string</code>
    * [.TABLE_INPUT_IMAGE_MAX_RES](#config46table95input95image95max95res) : <code>string</code>
    * [.TABLE_SAVE_STATE_DEBOUNCE_TIME](#config46table95save95state95debounce95time) : <code>string</code>

<a id="config46table95api95disconnect95timeout"></a>

### config.TABLE\_API\_DISCONNECT\_TIMEOUT : <code>string</code>
Timeout in ms for api disconnect.
Defaults to 15000.

**Kind**: static property of [<code>config</code>](#config)  
<a id="config46table95api95request95timeout"></a>

### config.TABLE\_API\_REQUEST\_TIMEOUT : <code>string</code>
Timeout in ms for a single api request.
Defaults to 30000.

**Kind**: static property of [<code>config</code>](#config)  
<a id="config46table95api95url"></a>

### config.TABLE\_API\_URL : <code>string</code>
URL to Websocket-API.
Defaults to /api.

**Kind**: static property of [<code>config</code>](#config)  
<a id="config46table95desktop95app95broadcast95image95interval"></a>

### config.TABLE\_DESKTOP\_APP\_BROADCAST\_IMAGE\_INTERVAL : <code>string</code>
Interval in ms for capturing screenshots during broadcast. 
Defaults to 15000.

**Kind**: static property of [<code>config</code>](#config)  
<a id="config46table95desktop95app95screenshot95max95res"></a>

### config.TABLE\_DESKTOP\_APP\_SCREENSHOT\_MAX\_RES : <code>string</code>
Max. width or height of captured screenshot in px. 
     TABLE_* If screenshot is too large, it will be scaled down til width and height
are less or equal to the specified value.
Defaults to 1200.

**Kind**: static property of [<code>config</code>](#config)  
<a id="config46table95html95contact95infos"></a>

### config.TABLE\_HTML\_CONTACT\_INFOS : <code>string</code>
HTML text for contact page.
Requires override.

**Kind**: static property of [<code>config</code>](#config)  
<a id="config46table95html95terms95of95service"></a>

### config.TABLE\_HTML\_TERMS\_OF\_SERVICE : <code>string</code>
HTML text for terms of service page.
Requires override.

**Kind**: static property of [<code>config</code>](#config)  
<a id="config46table95input95image95max95res"></a>

### config.TABLE\_INPUT\_IMAGE\_MAX\_RES : <code>string</code>
Maximum resolution (width or height) in px for an attached image.
Larger pictures get resized to this value.
Defaults to 1200.

**Kind**: static property of [<code>config</code>](#config)  
<a id="config46table95save95state95debounce95time"></a>

### config.TABLE\_SAVE\_STATE\_DEBOUNCE\_TIME : <code>string</code>
Time intervall in ms that state-save operations 
(save state to localStorage) can be performed at max rate.  
Defaults to 500.

**Kind**: static property of [<code>config</code>](#config)  


backend/src/config.js
// customConfig consists of all values of the process environment
// that start with the prefix "TABLE_"
const customConfig = Object.keys(process.env)
    .filter(key => key.startsWith('TABLE_'))    
    .reduce((customConfig, key) => {
        customConfig[key] = process.env[key];
        return customConfig;
    }, {});

/**
 * Configuration values.
 * IMPORTANT: Only string values are supported!
 */
const config = {
    /**
     * Passwort for access of (super) admin functions.
     * Should be totally secret!
     * Requires overwrite.
     * @type {string}
     */
    TABLE_ADMIN_PASSWORD: 'debug',
    /**
     * Logging-Level.
     * Valid values: 'debug', 'info', 'warn', 'error'.
     * Defaults to 'debug'.
     * @type {string}
     */
    TABLE_LOG_LEVEL: 'debug',
    /**
     * Base-URL of client. Needs to end with a slash.
     * Requires overwrite.
     * @type {string}
     */
    TABLE_BASE_URL: 'https://localhost:3000/',
    
    /**
     * Path to ssl-certificate.
     * Requires overwrite.
     * @type {string}
     */
    TABLE_SSL_CERT_PATH: './localhost.crt',
    /**
     * Path to ssl-certificates private key.
     * Requires overwrite.
     * @type {string}
     */
    TABLE_SSL_KEY_PATH: './localhost.key',

    /**
     * URL to LDAP-Server.
     * Requires overwrite.
     * @type {string}
     */
    TABLE_LDAP_URL: 'ldaps://ldap.example.com:636',
    /**
     * DN of search-account.
     * Requires overwrite.
     * @type {string}
     */
    TABLE_LDAP_DN: 'uid=abc123,...,dc=com',
    /**
     * Password of search-account.
     * Requires overwrite.
     * @type {string}
     */
    TABLE_LDAP_PASSWORD: 'password',
    /**
     * Search-base.
     * Requires overwrite.
     * @type {string}
     */
    TABLE_LDAP_BASE: 'ou=people,dc=example,dc=com',
    /**
     * LDAP entry attributes which are handled as email address.
     * Comma seperated list (e.g. "mail,mailShort,mailAlternative")
     * First entry should be the main attribute.
     * The main attribute is used to send mails to the user.
     * Defaults to 'mail'.
     * @type {string}
     */
    TABLE_LDAP_EMAIL_ATTRIBUTES: 'mail',
    /**
     * LDAP entry attribute which is handled as a persons name.
     * Defaults to 'cn'.
     * @type {string}
     */
    TABLE_LDAP_NAME_ATTRIBUTE: 'cn',
    /**
     * LDAP entry attribute which is handled as a persons title.
     * 
     * This setting is optional: if not set, no title is put in front of persons names.
     * @type {(string|undefined)}
     */
    TABLE_LDAP_TITLE_ATTRIBUTE: undefined,

    /**
     * Name of database to use.
     * Defaults to 'tableapp'.
     * @type {string}
     */
    TABLE_DB_NAME: 'tableapp',
    /**
     * URL to mongodb.
     * Defaults to local mongodb instance on default port.
     * @type {string}
     */
    TABLE_DB_URL: 'mongodb://localhost:27017',

    /**
     * Hostname of mail-server to connect to (via smtp).
     * Defaults to 'localhost'.
     * @type {string}
     */
    TABLE_MAIL_HOST: 'localhost',
    /**
     * Mail-address to send mails from.
     * Requires overwrite.
     * @type {string}
     */
    TABLE_MAIL_FROM: 'table@example.com',

    /**
     * Private key used for signing.
     * Requires overwrite.
     * @type {string}
     */
    TABLE_SESSION_TOKEN_PRIVATE_KEY: 'password',
    /**
     * Time span a sessionToken is valid after creation.
     * Defaults to '180d' (180 days).
     * 
     * Value is expressed in seconds or a string describing a time span. Eg: 60, "2 days", "10h", "7d"
     * @type {string}
     */
    TABLE_SESSION_TOKEN_VALID_FOR: '180d',

    /**
     * Time in minutes to keep screenshots available after last received screenshot/update.
     * Defaults to 15.
     * @type {string}
     */
    TABLE_EVENT_SCREENSHOTS_KEEP_FOR_MINUTES_AFTER_LAST_UPDATE: '15',
    /**
     * Determines how many screenshots should be preserved at a time.
     * Defaults to 3.
     * @type {string}
     */
    TABLE_EVENT_SCREENSHOTS_SAVE_LAST_COUNT: '3',

    ...customConfig
};

module.exports = config;
